paddle/majel/malloc.cc

```diff
@@ -0,0 +1,131 @@
+#include "malloc.h"
+#include <glog/logging.h>
+
+#ifndef PADDLE_ONLY_CPU
+#include <cuda_runtime.h>
+#endif
+
+#define CHECK_CUDA(cudaFunc)                                         \
+  do {                                                               \
+    cudaError_t cudaStat = cudaFunc;                                 \
+    CHECK_EQ(cudaSuccess, cudaStat) << "Cuda Error: "                \
+                                    << cudaGetErrorString(cudaStat); \
+  } while (0)
+
+namespace majel {
+namespace malloc {
+namespace detail {
+#ifndef PADDLE_ONLY_CPU
+const char* get_device_error_string() {
+  cudaError_t err = cudaGetLastError();
+  return cudaGetErrorString(err);
+}
+
+const char* get_device_error_string(size_t err) {
+  return cudaGetErrorString((cudaError_t)err);
+}
+
+void* malloc_device(size_t size) {
+  void* dest_d;
+
+  CHECK(size) << __func__ << ": the size for device memory is 0, please check.";
+  CHECK_CUDA(cudaMalloc((void**)&dest_d, size));
+
+  return dest_d;
+}
+
+void free_mem_device(void* dest_d) {
+  CHECK_NOTNULL(dest_d);
+
+  cudaError_t err = cudaFree(dest_d);
+  CHECK(cudaSuccess == err || cudaErrorCudartUnloading == err)
+      << get_device_error_string();
+}
+#endif
+
+class DefaultAllocator {
```

Comment:
Majel provides various memory management policies. Every memory management policy can be abstracted into a allocator class. Here, we just implement a simple one first, ```DefalutAllocator```.
```malloc``` is a global method which is responsible for memory allocation. ```malloc``` will choose a specific memory allocation policy.  And ```Allocation``` is a memory block handled by ```Array``` and will call ```malloc``` method.